listen/property-changes.js

```diff
@@ -70,8 +68,9 @@ PropertyChanges.prototype.getOwnPropertyChangeDescriptor = function (key) {
         });
     }
     var objectPropertyChangeDescriptors = this.__propertyChangeListeners__;
-    if (!object_owns.call(objectPropertyChangeDescriptors, key)) {
+    if (!objectPropertyChangeDescriptors.hasOwnProperty(key)) {
```

Comment:
To make these concerns real, just empathize with the developer who writes the JavaScript documentation browser that consumes a JSON blob like `{"Object": {"hasOwnProperty": {"returns": "whether property exists", "accepts": "a property name"}, "__proto__": {"property": "a reference to the Object.prototye"}, "toString": {"returns": "a string representation of the object"}}`.

These things do happen and waste real people’s time. Feel free to put it off if you need to take short-cuts to make a demo run fast.

That use case contains all the gotchas. If you can make that work, you’ll be fine. Object.create(null) still has a truthy "__proto__", but I vacillate about whether that’s worth worrying about (see Dict discussion; note that the code you’re improving does not already handle that case).

If you want to kill the correctness issue with fire, you have to prefix *all* entries in an object, regardless of whether it has a null proto.

If you want to do something slightly faster but equally correct, try using a null proto and special case "__proto__" as a property *of some other object*. Object.create may or may not be fast, so measure. You can also create null proto literals, but that deoptimizes the containing function in V8. `{"__proto__": null}`. You can isolate that with a small function `function nullObject() { return {"__proto__": null}`, but I don’t know how that compares to Object.create.

If Object.create turns out to be too slow, in addition to dealing with "__proto__", you have to deal with properties of Object.prototype. Varying levels of paranoia are appropriate when dealing with Object.prototype depending on whether you can rely on *everyone* to not add enumerable properties to the prototype. If you are blessed with an unaltered Object.prototype, there’s a fixed set of methods that come up as truthy on `object[key]` checks, for e.g., `key == "toString"`.

If the range (all possible values) of your map is all non-functions, you can distinguish them with `typeof object[key] !== "function"`. If the domain (all possible keys) of the map is guaranteed to exclude any JavaScript method names, e.g., all lower-case or all contain "http://", you don’t need to do anything special.

If none of that’s fast enough, you can consider having a separate Object to store special cases (testable with `if (emptyObject[key]`)  with a prefixed key.

Recall that the `in` operator both deoptimizes the containing function in V8, and traverses the entire prototype chain looking for enumerable keys. Allocating an array with the special `Object.keys(object)` is always preferable.

If you do not have any guarantees with regard to the Object.prototype, you have to use a null prototype for your map, or use a proper Map, or use a keys array and a corresponding values array.